src/mame/drivers/didact.cpp

```diff
@@ -457,34 +451,381 @@ void mp68a_state::machine_start()
+	required_device<cassette_image_device> m_cassette;
+	UINT8 *m_char_ptr;
+	UINT8 *m_vram;
+	UINT32 screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect);
+	virtual void machine_reset() override { m_maincpu->reset(); LOG(("--->%s()\n", FUNCNAME)); };
+	virtual void machine_start() override;
+	DECLARE_READ8_MEMBER( pia_r );
+	DECLARE_WRITE8_MEMBER( pia_w );
+	DECLARE_READ8_MEMBER( pia1_kbA_r );
+	DECLARE_WRITE8_MEMBER( pia1_kbA_w );
+	DECLARE_READ8_MEMBER( pia1_kbB_r );
+	DECLARE_WRITE8_MEMBER( pia1_kbB_w );
+protected:
+	required_device<pia6821_device> m_pia1;
+	required_device<pia6821_device> m_pia2;
+	required_ioport m_io_line5;
+	required_ioport m_io_line6;
+	required_ioport m_io_line7;
+	required_ioport m_io_line8;
+	required_ioport m_io_line9;
+	UINT8 m_line5;
+	UINT8 m_line6;
+	UINT8 m_line7;
+	UINT8 m_line8;
+	UINT8 m_line9;
+	UINT8 m_pia1_B;
+};
+
+void e100_state::machine_start()
+{
+	LOG(("%s()\n", FUNCNAME));
+	m_char_ptr  = memregion("chargen")->base();
+	m_vram      = (UINT8 *)m_videoram.target();
+}
+
+UINT32 e100_state::screen_update(screen_device &screen, bitmap_ind16 &bitmap, const rectangle &cliprect)
+{
+	int x, y;
+	int vramad;
+	UINT8 *chardata;
+	UINT8 charcode;
+
+	LOGSCREEN(("%s()\n", FUNCNAME));
+	vramad = 0;
+	for (int row = 0; row < 32 * 8; row += 8)
+	{
+		for (int col = 0; col < 32 * 8; col += 8)
+		{
+			/* look up the character data */
+			charcode = m_vram[vramad];
+			if (VERBOSE && charcode != 0x20 && charcode != 0) LOGSCREEN(("\n %c at X=%d Y=%d: ", charcode, col, row));
+			chardata = &m_char_ptr[(charcode * 8)];
+			/* plot the character */
+			for (y = 0; y < 8; y++)
+			{
+				if (VERBOSE && charcode != 0x20 && charcode != 0) LOGSCREEN(("\n  %02x: ", *chardata));
+				for (x = 0; x < 8; x++)
+				{
+					if (VERBOSE && charcode != 0x20 && charcode != 0) LOGSCREEN((" %02x: ", *chardata));
+					bitmap.pix16(row + y, col + x) = (*chardata & (1 << x)) ? 1 : 0;
+				}
+				chardata++;
+			}
+			vramad++;
+		}
+		if (VERBOSE && charcode != 0x20 && charcode != 0) LOGSCREEN(("\n"));
+	}
+
+	return 0;
+}
+
+/* PIA write - the Esselte 100 allows the PIA:s to be accessed simultaneously */
+WRITE8_MEMBER( e100_state::pia_w )
+{
+	LOG(("%s(%02x)\n", FUNCNAME, data));
+	if ((offset & 0x08) == 0x08)
+	{
+		LOG(("- PIA1\n"));
+		m_pia1->write(space, offset, data);
+	}
+	if ((offset & 0x10) == 0x10)
+	{
+		LOG(("- PIA2\n"));
+		m_pia2->write(space, offset, data);
+	}
+	if (VERBOSE && (offset & 0x18) == 0x18)
+	{
+		LOGCS(("- Dual device write access!\n"));
+	}
+	if (VERBOSE && (offset & 0x18) == 0x00)
+	{
+		logerror("- Funny write at offset %02x!\n", offset);
+	}
+}
+
+/* PIA read  - the Esselte 100 allows the PIA:s to be accessed simultaneously */
+READ8_MEMBER( e100_state::pia_r )
+{
+	UINT8 data = 0;
+
+	switch (offset & 0x18)
+	{
+	case 0x18: // read PIA1 and PIA2 at the same time, should really only happen for writes...
+		{
+			int data1 =  m_pia1->read(space, offset);
```

Comment:
the PIA read returns a `UINT8` - please don't play fast and loose with integer types.